### PR TITLE
[json-c] Add project

### PR DESCRIPTION
**Pull request by derwolfe**
This adds a fuzzer for the `json_tokener_parse_ex` function from [json-c](https://github.com/json-c/json-c), a popular library for working with JSON in C.

Hi @hawicz! Is this something you'd be interested in pursuing? I've not contributed to json-c in the past but I'd be happy to both work on adding more fuzzing targets and getting them pushed upstream :) 

I've been running this with `-detect_leaks=0` (to keep the fuzzer from crashing) as it has found several direct leaks locally. *Note*: This could be due to how the fuzzer is written and not the library itself.

### CODE REVIEW

File: projects/json-c/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+FROM gcr.io/oss-fuzz-base/base-builder
+MAINTAINER your@email.com
```
Comment (alex) [inline projects/json-c/Dockerfile]: Someone's email should be here.

File: projects/json-c/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+RUN apt-get update && apt-get install -y make autoconf automake libtool
+RUN git clone --depth 1 https://github.com/json-c/json-c.git json-c     # or use other version control
```
Comment (alex) [inline projects/json-c/Dockerfile]: Comment can go :-)

File: projects/json-c/build.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+# build project
+# e.g.
```
Comment (alex) [inline projects/json-c/build.sh]: same

File: projects/json-c/tokener_parse_ex.options [hunk 4]
```diff
@@ -0,0 +1,2 @@
+[libfuzzer]
+dict = tokener_parse_ex.dict
```
Comment (titanous) [inline projects/json-c/tokener_parse_ex.options]: This is unnecessary if the file is named `${fuzzer}.dict`

File: projects/json-c/tokener_parse_ex.dict [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+":"
+"e"
+"e+"
+"e-"
+"E"
+"E+"
+"E-"
+"\""
+"\\"
+"\/"
```
Comment (titanous) [inline projects/json-c/tokener_parse_ex.dict]: This dictionary does not appear to be valid, are you sure it's being used?

```
ParseDictionaryFile: error in line 16
		"\/"
```

File: projects/json-c/tokener_parse_ex_fuzzer.cc [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#include <stdint.h>
+
+#include <json.h>
+
+extern "C" int LLVMFuzzerTestOneInput(const char *data, size_t size) {
+  json_tokener *tok = json_tokener_new();
+  json_tokener_parse_ex(tok, data, size);
```
Comment (hawicz) [inline projects/json-c/tokener_parse_ex_fuzzer.cc]: If you're doing any leak checking you should free any generated json_object returned here with json_object_put().

**Comment by hawicz** [comment]
I'm not sure quite what you imply by "pursuing", but if you find bugs in json-c I'd be happy to apply any PRs you propose in order to fix them.
If you're seeing leaks, even after properly releasing (with json_object_put) parsed objects that's definitely something I'd be interested in getting fixed.

File: projects/json-c/tokener_parse_ex_fuzzer.cc (hunk 6 above)
Comment (derwolfe) [reply projects/json-c/tokener_parse_ex_fuzzer.cc]: thank you - I'll do this.

File: projects/json-c/tokener_parse_ex.dict (hunk 5 above)
Comment (derwolfe) [reply projects/json-c/tokener_parse_ex.dict]: I'm not and given that you've found this, I'd expect that it isn't. How did you notice this error?

**Comment by derwolfe** [comment]
> Hi @hawicz! Is this something you'd be interested in pursuing?

Sorry, that was imprecise of me :) By pursuing, I meant to ask if you'd be interested in getting the project integrated into OSS-Fuzz.

File: projects/json-c/tokener_parse_ex.dict (hunk 5 above)
Comment (titanous) [reply projects/json-c/tokener_parse_ex.dict]: I copied it for #1119. Look for `-dict=` in the command line that is printed at the beginning of the `run_fuzzer` output.

File: projects/json-c/tokener_parse_ex_fuzzer.cc (hunk 6 above)
Comment (derwolfe) [reply projects/json-c/tokener_parse_ex_fuzzer.cc]: That fixed my leaks - thanks.

**Comment by hawicz** [comment]
I have no idea whether json-c satisfies the "must have a significant user base and/or be critical to the global IT infrastructure" requirement.  Given the number of places I've happened to see it used, I suspect it may, but I have not explicitly collected any stats.
That aside, I do know that the developer base is a bit small, and though I'd be happy to consider additional qualified people to join as "official" developers, I don't want to mislead anyone into thinking that including the project here is guaranteed to get a particularly fast response to problems found from myself or the other three people currently listed on the json-c project.

File: projects/json-c/tokener_parse_ex.dict (hunk 5 above)
Comment (derwolfe) [reply projects/json-c/tokener_parse_ex.dict]: :+1: thanks

**Comment by inferno-chromium** [comment]
@hawicz, @derwolfe - Can you please comment on who are the consumers of json-c, we need to understand how widely is this project used.

**Comment by derwolfe** [comment]
@inferno-chromium I found it when looking through [https://brew.sh/analytics/install/](homebrew project rankings).

**Comment by inferno-chromium** [comment]
@hawicz - can you please expand more on the last comment from @derwolfe, i want to make sure that this is used in production in some places.

**Comment by derwolfe** [comment]
A quick semi-imprecise check shows that the following debian packages depend on this

```
apt-cache rdepends libjson-c2
libjson-c2
Reverse Depends:
  libjson-c-dev
  libfwts1
  google-compute-engine-oslogin
  fwts
  upstart
  strongswan-tnc-server
  libpulse0
  libclamav7
  zmap
  syslog-ng-mod-json
  sagan
  rt-app
  riemann-c-client
  ptask
  psensor-server
  psensor
  postgresql-9.5-postgis-2.2
  pd-purest-json
  ntopng
  nfstrace
  newsbeuter
  mypaint
  libykpers-1-1
  libxr1
  libubuntu-location-service3
  libu2f-server0
  libu2f-host0
  libu1db1
  libshogun16
  libpiano0
  libndpi-bin
  liblwgeom-2.2-5
  liblognorm2
  libgirara-gtk3-2
  libgdcm2.6
  libfwts1
  kamailio-kazoo-modules
  kamailio-json-modules
  hubicfuse
  hhvm
  health-check
  guake-indicator
  fwts
  fastd
  bti
  upstart
  strongswan-tnc-server
  rsyslog
  libpulse0
  libjson0
  libjson-c2-dbg
  libclamav7
```

On fedora:

```
[root@dd0cfc03213e /]# dnf repoquery -q --whatrequires json-c
R-shogun-0:6.0.0-5.fc27.x86_64
abrt-0:2.10.4-3.fc27.x86_64
abrt-0:2.10.5-1.fc27.x86_64
abrt-plugin-bodhi-0:2.10.4-3.fc27.x86_64
abrt-plugin-bodhi-0:2.10.5-1.fc27.x86_64
bluez-0:5.47-2.fc27.x86_64
bluez-0:5.48-1.fc27.x86_64
bti-0:034-10.fc27.x86_64
daxctl-0:58.2-1.fc27.x86_64
daxctl-0:58.4-1.fc27.x86_64
fastd-0:18-4.fc27.x86_64
filezilla-0:3.30.0-1.fc27.x86_64
freeradius-rest-0:3.0.15-3.fc27.x86_64
fwts-0:17.09.00-1.fc27.x86_64
gdal-0:2.1.4-5.fc27.x86_64
gdal-java-0:2.1.4-5.fc27.x86_64
gdal-libs-0:2.1.4-5.fc27.i686
gdal-libs-0:2.1.4-5.fc27.x86_64
gdcm-0:2.6.5-13.fc27.i686
gdcm-0:2.6.5-13.fc27.x86_64
gfal2-0:2.14.2-3.fc27.i686
gfal2-0:2.14.2-3.fc27.x86_64
gfal2-plugin-dcap-0:2.14.2-3.fc27.x86_64
gfal2-plugin-file-0:2.14.2-3.fc27.x86_64
gfal2-plugin-gridftp-0:2.14.2-3.fc27.i686
gfal2-plugin-gridftp-0:2.14.2-3.fc27.x86_64
gfal2-plugin-http-0:2.14.2-3.fc27.i686
gfal2-plugin-http-0:2.14.2-3.fc27.x86_64
gfal2-plugin-lfc-0:2.14.2-3.fc27.x86_64
gfal2-plugin-mock-0:2.14.2-3.fc27.x86_64
gfal2-plugin-rfio-0:2.14.2-3.fc27.x86_64
gfal2-plugin-sftp-0:2.14.2-3.fc27.x86_64
gfal2-plugin-srm-0:2.14.2-3.fc27.i686
gfal2-plugin-srm-0:2.14.2-3.fc27.x86_64
gfal2-plugin-xrootd-0:2.14.2-3.fc27.x86_64
girara-0:0.2.7-4.fc27.i686
girara-0:0.2.7-4.fc27.x86_64
gluster-block-0:0.2.1-2.fc27.x86_64
gluster-block-0:0.3-1.fc27.x86_64
java-shogun-0:6.0.0-5.fc27.x86_64
json-c-devel-0:0.12.1-4.fc27.i686
json-c-devel-0:0.12.1-4.fc27.x86_64
json-c-devel-0:0.12.1-5.fc27.i686
json-c-devel-0:0.12.1-5.fc27.x86_64
libdmmp-0:0.7.1-6.git847cc43.fc27.i686
libdmmp-0:0.7.1-6.git847cc43.fc27.x86_64
libdmmp-0:0.7.1-9.git847cc43.fc27.i686
libdmmp-0:0.7.1-9.git847cc43.fc27.x86_64
libmypaint-0:1.3.0-4.fc27.i686
libmypaint-0:1.3.0-4.fc27.x86_64
libreport-plugin-bugzilla-0:2.9.2-1.fc27.x86_64
libreport-plugin-bugzilla-0:2.9.3-1.fc27.x86_64
libreport-plugin-kerneloops-0:2.9.2-1.fc27.x86_64
libreport-plugin-kerneloops-0:2.9.3-1.fc27.x86_64
libreport-plugin-mantisbt-0:2.9.2-1.fc27.x86_64
libreport-plugin-mantisbt-0:2.9.3-1.fc27.x86_64
libreport-plugin-reportuploader-0:2.9.2-1.fc27.x86_64
libreport-plugin-reportuploader-0:2.9.3-1.fc27.x86_64
libreport-plugin-rhtsupport-0:2.9.2-1.fc27.x86_64
libreport-plugin-rhtsupport-0:2.9.3-1.fc27.x86_64
libreport-plugin-ureport-0:2.9.2-1.fc27.x86_64
libreport-plugin-ureport-0:2.9.3-1.fc27.x86_64
libreport-web-0:2.9.2-1.fc27.i686
libreport-web-0:2.9.2-1.fc27.x86_64
libreport-web-0:2.9.3-1.fc27.i686
libreport-web-0:2.9.3-1.fc27.x86_64
libstorj-0:1.0.2-2.fc27.i686
libstorj-0:1.0.2-2.fc27.x86_64
libu2f-host-0:1.1.3-3.fc27.i686
libu2f-host-0:1.1.3-3.fc27.x86_64
libu2f-server-0:1.0.1-10.fc27.i686
libu2f-server-0:1.0.1-10.fc27.x86_64
libu2f-server-0:1.0.1-7.fc27.i686
libu2f-server-0:1.0.1-7.fc27.x86_64
libverto-jsonrpc-0:0.1.0-16.fc27.i686
libverto-jsonrpc-0:0.1.0-16.fc27.x86_64
libverto-jsonrpc-0:0.1.0-9.fc21.i686
libverto-jsonrpc-0:0.1.0-9.fc21.x86_64
libvmi-0:0.11.0-10.20170706gite919365.fc27.x86_64
libvmi-utils-0:0.11.0-10.20170706gite919365.fc27.x86_64
lua-shogun-0:6.0.0-5.fc27.x86_64
mod_lcgdm_dav-0:0.18.2-2.fc27.x86_64
mod_lcgdm_dav-0:0.19.0-1.fc27.x86_64
mono-shogun-0:6.0.0-5.fc27.x86_64
mypaint-0:1.1.0-15.fc27.x86_64
mypaint-0:1.2.1-3.fc27.i686
mypaint-0:1.2.1-3.fc27.x86_64
ndctl-0:58.2-1.fc27.x86_64
ndctl-0:58.4-1.fc27.x86_64
newsbeuter-0:2.9-2.fc24.x86_64
newsbeuter-0:2.9-6.fc27.x86_64
octave-shogun-0:6.0.0-5.fc27.x86_64
opensips-json-0:2.2.3-9.fc27.x86_64
opensips-json-0:2.2.5-2.fc27.x86_64
postgis-0:2.3.3-1.fc27.i686
postgis-0:2.3.3-1.fc27.x86_64
python2-shogun-0:6.0.0-5.fc27.x86_64
python3-shogun-0:6.0.0-5.fc27.x86_64
riemann-c-client-0:1.9.0-7.fc27.i686
riemann-c-client-0:1.9.0-7.fc27.x86_64
ruby-shogun-0:6.0.0-5.fc27.x86_64
shogun-0:6.0.0-5.fc27.i686
shogun-0:6.0.0-5.fc27.x86_64
strongswan-tnc-imcvs-0:5.5.2-1.fc27.x86_64
strongswan-tnc-imcvs-0:5.6.0-1.fc27.x86_64
sway-0:0.14.0-2.fc27.x86_64
sway-0:0.15.1-1.fc27.x86_64
syslog-ng-0:3.10.1-3.fc27.i686
syslog-ng-0:3.10.1-3.fc27.x86_64
systemtap-runtime-0:3.2-0.20170516gitc67d8f274b21.fc27.x86_64
systemtap-runtime-0:3.2-2.fc27.x86_64
tlog-0:3-3.fc27.i686
tlog-0:3-3.fc27.x86_64
zmap-0:2.1.1-4.fc27.x86_64
```

File: projects/json-c/tokener_parse_ex_fuzzer.cc [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#include <stdint.h>
+
+#include <json.h>
+
+extern "C" int LLVMFuzzerTestOneInput(const char *data, size_t size) {
+  json_tokener *tok = json_tokener_new();
+  json_object *obj;
```
Comment (Dor1s) [inline projects/json-c/tokener_parse_ex_fuzzer.cc]: why not call `json_tokener_parse_ex` from line 9 right here?

File: projects/json-c/project.yaml [hunk 8]
```diff
@@ -0,0 +1,4 @@
+homepage: "https://json-c.github.io/json-c/"
+primary_contact: "<primary_contact_email>"
```
Comment (Dor1s) [inline projects/json-c/project.yaml]: who will be a primary contact, @hawicz ?

File: projects/json-c/tokener_parse_ex_fuzzer.cc (hunk 7 above)
Comment (derwolfe) [reply projects/json-c/tokener_parse_ex_fuzzer.cc]: No reason at all. Will change.

**Comment by inferno-chromium** [comment]
Can you please put email for primary_contact and then we can merge. hawicz@ what email would you like to put here or want to recommend any other maintainers (see https://github.com/google/oss-fuzz#accepting-new-projects) ?

File: projects/json-c/tokener_parse_ex_fuzzer.cc [hunk 9]
```diff
@@ -0,0 +1,12 @@
+#include <stdint.h>
+
+#include <json.h>
+
+extern "C" int LLVMFuzzerTestOneInput(const char *data, size_t size) {
```
Comment (derwolfe) [inline projects/json-c/tokener_parse_ex_fuzzer.cc]: This signature is wrong; ubsan will fail. I'm going to modify this to take `unsigned char` and use a `reinterpret_cast`

File: projects/json-c/project.yaml (hunk 8 above)
Comment (hawicz) [reply projects/json-c/project.yaml]: You can use erh+git@nimenees.com

**Comment by hawicz** [comment]
For consumers of json-c, besides the packages that depend on it, I happen to know that it's been used in the software for a network HSM product, and at least at some point in the past it was used within the credit reporting system at Trans Union, but whether that is still the case I couldn't say.

**Pull request merged**